fix: fix editor progress position when enabled pinnedTabsOnSeparateRow

harbin1053020115 opened the pull request:
<!-- Thank you for submitting a Pull Request. Please:
* Read our Pull Request guidelines:
  https://github.com/microsoft/vscode/wiki/How-to-Contribute#pull-requests
* Associate an issue with the Pull Request.
* Ensure that the code is up-to-date with the `main` branch.
* Include a description of the proposed changes and how to test them.
-->
To #195313

gjsjohnmurray commented:
/assign @benibenj

benibenj commented on src/vs/workbench/browser/media/part.css:
```diff
@@ -90,7 +90,7 @@
 	position: absolute;
 	left: 0;
-	top: 33px; /* at the bottom of the 35px height title container */
+	top: var(--editor-group-tab-total-height);
```
The progress bar is not only used for the editor group title bar, but also in many other areas so we can't just change this value. It could however be overwritten just for the editor title bar for example [here](https://github.com/microsoft/vscode/blob/9dbe48372d1b3ac1abc92ff8f319d1edcefeaaef/src/vs/workbench/browser/parts/editor/media/editortitlecontrol.css).

benibenj replied:
I like that you introduced a --editor-group-tab-total-height variable, however, you still need to remove 2px due to the height of the progress bar.

benibenj reviewed (changes requested):
Thanks for submitting the bug and creating a PR! I left some comments to your changes

harbin1053020115 commented:
@benibenj You are right. The first commit won't work. Even `pinnedTabsonSeparateRow` setting is activated, there will be just one tab row if where is no pinned tabs opened.
And, I found it's not a good way to update tab height dynamic using css variable, we need to conside many scenarios:
1. When open/close/pinned editors, we need to update height according to opened tabs height
2. When moving opened editors in editor title, we need to update height according to opened tabs height
3. When `pinnedTabsonSeparateRow`/`editorTabHeight` setting is changed, we need to update height
So, I have changed the way to realize, I moved the editor progress bar into titleContainer, and make the progress bar at the bottom of titleContainer

harbin1053020115 commented:
Below is the result of progress bar:
1. normal: 
![image](https://github.com/microsoft/vscode/assets/2869750/c23b97a7-61fd-4f57-baec-e8a3affb4695)
2. `window.density.editorTabHeight` setting changed to compact:
![image](https://github.com/microsoft/vscode/assets/2869750/1c1e1153-6661-481d-8ab4-8d6bfbeedab3)
3. `breadcrumbs.enabled` enabled:
![image](https://github.com/microsoft/vscode/assets/2869750/8e68030b-5e00-4901-ad59-301c5e397890)
4. `pinnedTabsonSeparateRow` enabled:
![image](https://github.com/microsoft/vscode/assets/2869750/0238daad-9d73-44e1-ad2a-d517dfd6a1d7)

But it seems the progress bar' height is different

bpasero commented on src/vs/workbench/browser/parts/editor/media/editortitlecontrol.css:
```diff
@@ -45,3 +45,8 @@
 }
+
+.monaco-workbench .part.editor > .content .title .monaco-progress-container {
+	top: unset;
```
What is this change about? Can it have a comment to explain?

harbin1053020115 replied:
This is reset styles about absolute position which is defined in src/vs/workbench/browser/media/part.css.
I have add a comment too. See the latest commit.

benibenj commented:
I think we should still set the progress bar from the top as we did before. Your approach causes it to be placed underneath the breadcrumbs which did not used to be the case, and so we don't want to change that. 

This is my recommendation:

We already know the height of the editor title control. It can be accessed in 
[editorGroupView.ts](https://github.com/harbin1053020115/vscode/blob/1ef2da773206fca8582fa2c32dabbc6e319dd75c/src/vs/workbench/browser/parts/editor/editorGroupView.ts#L776-L778) or [editorTitleControl.ts](https://github.com/harbin1053020115/vscode/blob/1ef2da773206fca8582fa2c32dabbc6e319dd75c/src/vs/workbench/browser/parts/editor/editorTitleControl.ts#L209-L217). Like in your first approach, you can introduce a CSS variable, and overwrite the absolute top position of the progress bar in editorgroupview.css or editorTitleControl.css (depending on what level you operate on). The variable could be called `--editor-group-tabs-height` as we already have `--editor-group-tab-height`. I wouldn't call it title height as we do not want to include the breadcrumbs height.

The last thing to do now is update the value when there are changes to the layout, these could be:
- changing `window.density.editorTabHeight` setting
- changing `workbench.editor.pinnedTabsOnSeparateRow` setting
- changing `workbench.editor.wrapTabs` setting
- When opening, closing and moving an editor
- When sticking and unsticking (known as pinning and unpinning) an editor

The last two are specific to having some settings turned on, however, we can start off by updating the height every time as it is not an expensive operation and future proof. But we might change this in the future.

harbin1053020115 commented:
> I think we should still set the progress bar from the top as we did before. Your approach causes it to be placed underneath the breadcrumbs which did not used to be the case, and so we don't want to change that.
> 
> This is my recommendation:
> 
> We already know the height of the editor title control. It can be accessed in [editorTitleControl.ts](https://github.com/harbin1053020115/vscode/blob/1ef2da773206fca8582fa2c32dabbc6e319dd75c/src/vs/workbench/browser/parts/editor/editorTitleControl.ts#L209-L217). Like in your first approach, you can introduce a CSS variable, and overwrite the absolute top position of the progress bar in editorgroupview.css (This is where the progress bar lives). The variable could be called `--editor-group-tabs-height` as we already have `--editor-group-tab-height`. I wouldn't call it title height as we do not want to include the breadcrumbs height.
> 
> The last thing to do now is update the value when there are changes to the layout, these could be:
> 
> * changing `window.density.editorTabHeight` setting
> * changing `workbench.editor.pinnedTabsOnSeparateRow` setting
> * changing `workbench.editor.wrapTabs` setting
> * On `layout()` (in case of resizing the editor for example) when using wrapped tabs
> * When opening, closing and moving an editor
> * When sticking and unsticking (known as pinning and unpinning) an editor
> 
> The last two are specific to having some settings turned on, however, we can start off by updating the height every time as it is not an expensive operation and future proof. But we might change this in the future.

Got it. Let me submit a change.

benibenj commented:
@harbin1053020115 The change looks great! Thanks very much. 
I removed the method you added as we can update the variable in the editorGroupView when we call the layout method.

harbin1053020115 commented:
> @harbin1053020115 The change looks great! Thanks very much. I removed the method you added as we can update the variable in the editorGroupView when we call the layout method.

Great~Moving to layout to reset css variable is a better way.
